core/src/main/java/org/eclipse/hono/util/HonoProtonHelper.java

```diff
@@ -164,4 +165,44 @@ public static void setDefaultCloseHandler(final ProtonSession session) {
         }
         return result;
     }
+
+    /**
+     * Close the passed link and later call its {@link ProtonLink#free()} method in an attached closeHandler.
+     * <p>
+     * Additionally supervise if this handler is called in the number of {@link Constants#DEFAULT_FREE_LINK_AFTER_CLOSE_INTERVAL_MILLIS}
+     * milliseconds. If not, the timer is triggered and the link will be freed inside the timer code.
+     *
+     * @param context The context to run the timer code on.
+     * @param link The link to close and free.
+     * @param closeHandler The handler to invoke after the link was freed.
+     */
+    public static void closeAndFree(final Context context, final ProtonLink link, final Future<Void> closeHandler) {
+        final AtomicLong closeSenderTimerId = new AtomicLong(-1);
+
+        final Handler<Void> freeLinkHandler = v -> {
+            closeHandler.complete();
+            link.free();
+        };
+
+        // if sender gets remote peer detach close -> complete senderCloseHandler
+        link.closeHandler(closeAttempt -> {
+            context.owner().cancelTimer(closeSenderTimerId.get());
+            freeLinkHandler.handle(null);
+        });
+
+        if (link.isOpen()) {
+            // wait for peer's detach frame to trigger the handler
+            link.close();
+            // but add a timer to free resources if peer did not detach
+            closeSenderTimerId.set(context.owner().setTimer(Constants.DEFAULT_FREE_LINK_AFTER_CLOSE_INTERVAL_MILLIS, timerId -> {
```

Comment:
IMHO we should start the timer *before* sending our detach frame in order to make sure that the timer ID is set when the remote detach frame arrives ...